pkg/util/cephcmds.go

```diff
@@ -0,0 +1,364 @@
+/*
+Copyright 2019 The Ceph-CSI Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package util
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"k8s.io/klog"
+	"os"
+	"os/exec"
+	"strings"
+)
+
+func execCommand(program string, args ...string) (stdout, stderr []byte, err error) {
+	var (
+		cmd       = exec.Command(program, args...) // nolint: gosec
+		stdoutBuf bytes.Buffer
+		stderrBuf bytes.Buffer
+	)
+
+	cmd.Stdout = &stdoutBuf
+	cmd.Stderr = &stderrBuf
+
+	if err := cmd.Run(); err != nil {
+		return stdoutBuf.Bytes(), stderrBuf.Bytes(), fmt.Errorf("an error (%v)"+
+			" occurred while running %s", err, program)
+	}
+
+	return stdoutBuf.Bytes(), nil, nil
+}
+
+// CephStoragePoolSummary strongly typed JSON spec for osd ls pools output
+type CephStoragePoolSummary struct {
+	Name   string `json:"poolname"`
+	Number int    `json:"poolnum"`
+}
+
+// GetPoolID searches a list of pools in a cluster and returns the ID of the pool that matches
+// the passed in poolName parameter
+func GetPoolID(monitors string, adminID string, key string, poolName string) (uint32, error) {
+	// ceph <options> -f json osd lspools
+	// JSON out: [{"poolnum":<int>,"poolname":<string>}]
+
+	stdout, _, err := execCommand(
+		"ceph",
+		"-m", monitors,
+		"--id", adminID,
+		"--key="+key,
+		"-c", CephConfigPath,
+		"-f", "json",
+		"osd", "lspools")
+	if err != nil {
+		klog.Errorf("failed getting pool list from cluster (%s)", err)
+		return 0, err
+	}
+
+	var pools []CephStoragePoolSummary
+	err = json.Unmarshal(stdout, &pools)
+	if err != nil {
+		klog.Errorf("failed to parse JSON output of pool list from cluster (%s)", err)
+		return 0, fmt.Errorf("unmarshal failed: %+v.  raw buffer response: %s", err, string(stdout))
+	}
+
+	for _, p := range pools {
+		if poolName == p.Name {
+			return uint32(p.Number), nil
```

Comment:
question: can we make struct filed as unit32 to avoid type conversion?

Reply:
Do you mean the `Number` field at line 51? I think so, -ve pool IDs should not be valid anyway, so int should not be required there.

Reply:
Correct, negative pool numbers are not valid (but -1 is used internally in the APIs to indicate pool-not-found / pool-not-valid versus the "0" returned below in the error case which is a possibly valid pool id).

Reply:
So, making that uint32 would be fine, right?

Also, the line below has to return a value (in this case 0), but we should possibly rely on the go idiom to check errors before using return values, and hence the 0 below should be fine?

Reply:
It's actually a ```int64_t``` internally (not that I expect 63bits of a pool number) [1]. 

[1] https://github.com/ceph/ceph/blob/master/src/include/rados/librados.h#L839